### PR TITLE
adds all properties passed to component through helper method

### PR DESCRIPTION
**Pull request by gkemmey**
Seems like it'd be nice to get access to all the properties given to a component.

#### InputComponent

```rb
module InputComponent
  extend ComponentHelper

  property :f, required: true
  property :name, required: true

  def options
    @properties.slice(:f, :name).merge({
      class: (Array(properties[:class]) + ["text-input"]).join(' ')
    })
  end
end
```

#### _input.html.erb

```html+ruby
<%= @f.text_field @name, options %>
```

#### _new_todo.html.erb
```html+ruby
<%= form_with model: Todo.new do |f| %>
  <%= component 'input' f: f, name: :tile, id: "new-todo",
                                           placeholder: "What needs to be done?",
                                           autofocus: true,
                                           autocomplete: "off" %>
<% end %>
```

`InputComponent` becomes a wrapper around a call to `text_field` that you could apply some stylings to or whatever.

### Testing

I wrote some tests that involved adding another app under `fixtures` called `test_app` (hence all the files). On that, I created a couple components I could use to test expected rendering in a `ActiveSupport::ViewTest`.

Looks like some of the testing process is in limbo in other PRs, so happy to adjust if need be.

### CODE REVIEW

File: lib/komponent/komponent_helper.rb [hunk 1]
```diff
@@ -40,6 +40,7 @@ def component(component, locals = {}, &block)
         instance_variable_set(:"@#{name}", locals[name])
       end
 
+      instance_variable_set(:@properties, locals)
```
Comment (gkemmey) [inline lib/komponent/komponent_helper.rb]: Relevant non-test change.

**Comment by Spone** [comment]
Hi @gkemmey thanks for the PR! We'll review this with the team.

First thing that comes to my mind: I think we would need to "blacklist" the use of `properties` as a variable passed to the `component` helper, to make sure there is no risk of collision.

In other terms, we would raise an exception in case you write:

`= component "input", properties: {}`

so it doesn't conflict with using `@properties`.

Another approach would be to access properties using a `properties` method, instead of a `@properties` instance variable.

I think I prefer this second approach. What do you think?

File: test/komponent/komponent_helper_test.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+class KomponentHelperTest < ActionView::TestCase
+  def test_helper_renders_makes_locals_available_as_instance_variables
+    assert_equal %(<div class="world">🌎</div>), component('world', world: "🌎").chomp
+  end
+
+  def test_helper_makes_all_properties_accessible
+    assert_equal \
+      %(<div class="all">🌎 😎</div>),
+      component('all', world: "🌎", sunglasses: "😎").chomp
```
Comment (Spone) [inline test/komponent/komponent_helper_test.rb]: ahah, love the emojis :nerd_face:

**Comment by gkemmey** [comment]
Ah, makes sense. I agree that's nicer. So are you thinking

```rb
define_singleton_method(:properties) { locals }
```

would work instead?

**Comment by florentferry** [comment]
Hello @gkemmey,

Great idea to make the same variables available through `properties` or via instance variables. The method `properties` seems good to me.

Can you reuse the existing `fixtures/my_app` to make your tests instead of creating a new one ?  I don't care between `cucumber`, `rspec` or `test-unit`, take what you like to make your tests. The 2 related PRs can give you inspiration for adapting `fixtures/my_app` in order to make tests working.

We should use the foundation you made to implement other tests for `component_helper`.

#63 #32

**Comment by gkemmey** [comment]
@florentferry, @Spone  Thanks for reviewing!

Think this is fixed, let me know if anything else needs to change.

**Comment by florentferry** [comment]
Thanks for the changes. Good to merge for me, just need a mention about the variables accessing via properties in https://github.com/komposable/komponent#passing-variables, and a addition to changelog.

**Comment by Spone** [comment]
Thanks for the changes, ready to merge!

**Pull request merged**